lib/cn/csched_sp3.c

```diff
@@ -1134,37 +1151,105 @@ sp3_process_ingest(struct sp3 *sp)
 static void
 sp3_process_dirtylist(struct sp3 *sp)
 {
-    struct cn_tree *tree;
+    struct cn_tree *tree, *tree_next;
 
-    list_for_each_entry(tree, &sp->mon_tlist, ct_sched.sp3t.spt_tlink) {
+    list_for_each_entry_safe(tree, tree_next, &sp->mon_tlist, ct_sched.sp3t.spt_tlink) {
         struct sp3_tree *spt = tree2spt(tree);
-        struct sp3_node *spn;
+        struct sp3_node *spn, *spn_next;
+        struct list_head joined;
+        uint ndirty = 0;
         void *lock;
 
+        if (atomic_read(&sp->sp_dirty_count) == 0)
+            return;
```

Comment:
I use this technique elsewhere to avoid having to scan all the trees on each iteration of the monitor loop.  Since we have several functions that would otherwise do that this approach eliminates a lot of unnecessary work.
Probably the optimal way to reduce work is to have lists of trees that are ready for specific work, not sure if that's possible.